File: python_solutions/day_3.py
from itertools import islice
from typing import List, Tuple


def get_item_priority(item: str) -> int:
    item_ascii_code = ord(item)
    if item.upper() == item:
        return item_ascii_code - ord("A") + 27
    else:
        return item_ascii_code - ord("a") + 1


class Elf:
    def __init__(self, coded_line: str) -> None:
        self._first_compartment_contents = coded_line[: len(coded_line) // 2]
        self._second_compartment_contents = coded_line[len(coded_line) // 2 :]

    @property
    def held_items(self) -> str:
        return self._first_compartment_contents + self._second_compartment_contents

    def has(self, item: str) -> bool:
        return item in self.held_items

    def find_duplicated_item(self) -> str:
        for item in self._first_compartment_contents:
            if item in self._second_compartment_contents:
                return item
        raise IOError("No dubplicated item found")


class ElfGroup:
    def __init__(self, elves: Tuple[Elf, Elf, Elf]) -> None:
        self._elves = elves

    @property
    def elves(self) -> Tuple[Elf, Elf, Elf]:
        return self._elves

    def find_common_item(self) -> str:
        for item in self._elves[0].held_items:
            if self._elves[1].has(item) and self._elves[2].has(item):
                return item
        raise IOError("No common item found")


def group_elves() -> List[ElfGroup]:
    with open("../inputs/day_3.txt", "r") as f:
        elf_groups = []
        next_3_lines = list(islice(f, 3))
        while next_3_lines:
            elf_groups.append(
                ElfGroup(
                    (Elf(next_3_lines[0]), Elf(next_3_lines[1]), Elf(next_3_lines[2]))
                )
            )
            next_3_lines = list(islice(f, 3))
        return elf_groups


if __name__ == "__main__":

    duplicated_items_total = 0
    common_items_total = 0

    elf_groups = group_elves()
    for elf_group in elf_groups:
        common_items_total += get_item_priority(elf_group.find_common_item())
        for elf in elf_group.elves:
            duplicated_items_total += get_item_priority(elf.find_duplicated_item())

    print(f"Total priorities of duplicated items: {duplicated_items_total}")
    print(f"Total priorities of common items: {common_items_total}")

File: cpp_solutions/day_3.cpp
#include "day_3.hpp"
#include <cctype>
#include <iostream>
#include <fstream>
#include <stdexcept>
#include <functional>

int getItemPriority(char item)
{
    int asciiCode = (int)item;
    int firstLowerCode = (int)'a';
    int firstCapitalCode = (int)'A';

    if (isupper(item)) {
        return asciiCode - firstCapitalCode + 27;
    } else {
        return asciiCode - firstLowerCode + 1;
    };
}

std::string Elf::getHeldItems()
{
    return firstCompartment + secondCompartment;
}

bool Elf::has(char item)
{
    return (getHeldItems().find(item) != std::string::npos);
}

char Elf::findDuplicatedItem()
{
    for (int i = 0; i < firstCompartment.length(); i++) {
        char item = firstCompartment.at(i);
        if (secondCompartment.find(item) != std::string::npos) {
                return item;
            }
      }
    throw std::runtime_error("Cannot find duplicated item");
}

std::string Elf::decodeFirstCompartment(std::string codedItems)
{
    return codedItems.substr(0, codedItems.length() / 2);
}

std::string Elf::decodeSecondCompartment(std::string codedItems)
{
    return codedItems.substr(codedItems.length() / 2, codedItems.length() / 2);
}

char ElfGroup::findCommonItem()
{
    int numHeldItems = elves[0].getHeldItems().length();
    if (numHeldItems == 0) {throw std::runtime_error("All elves should have some items.");};
    try {
        for (int i = 0; i < elves[0].getHeldItems().length(); i++) {
            char item = elves[0].getHeldItems().at(i);
            if (elves[1].has(item) and elves[2].has(item)) {
                return item;
            }
        }
    } catch (const std::out_of_range&) {
        std::cout << "OUT OF RANGE ERROR!!!!!!!!!" << std::endl;
        std::cout << elves[0].getHeldItems() << std::endl;
    };
    throw std::runtime_error("Cannot find common item");

}

std::vector<ElfGroup> groupElves()
{
    std::ifstream file("../inputs/day_3.txt");
    std::vector<ElfGroup> elfGroups;

    while (true) {
        std::string lines[3];
        std::getline(file, lines[0]);
        std::getline(file, lines[1]);
        if (std::getline(file, lines[2])) {
            Elf elves[3] = {
                {std::ref(lines[0])},
                {std::ref(lines[1])},
                {std::ref(lines[2])}
            };
            ElfGroup elfGroup(elves);
            elfGroups.push_back(elfGroup);
        } else {
            break;
        }
    }

    return elfGroups;
}

int main()
{
    int duplicatedItemsTotal = 0;
    int commonItemsTotal = 0;

    std::vector<ElfGroup> elfGroups = groupElves();
    
    for (ElfGroup & elfGroup : elfGroups) {
        commonItemsTotal += getItemPriority(elfGroup.findCommonItem());
        for (int i = 0; i < 3; i++) {
            duplicatedItemsTotal += getItemPriority(
                elfGroup.elves[i].findDuplicatedItem()
                );
        }
    }

    std::cout << "Total priorities of duplicated items: " << duplicatedItemsTotal << std::endl;
    std::cout << "Total priorities of common items: " << commonItemsTotal << std::endl;
}